spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/functions/BucketFunction.java

```diff
@@ -0,0 +1,303 @@
+  private static final int NUM_BUCKETS_ORDINAL = 0;
+  private static final int VALUE_ORDINAL = 1;
+  private static final Set<DataType> SUPPORTED_NUM_BUCKETS_TYPES =
+      ImmutableSet.of(DataTypes.ByteType, DataTypes.ShortType, DataTypes.IntegerType);
+
+  @Override
+  public BoundFunction bind(StructType inputType) {
+    if (inputType.size() != 2) {
+      throw new UnsupportedOperationException(
+          "Wrong number of inputs (expected numBuckets and value)");
+    }
+
+    StructField numBucketsField = inputType.fields()[NUM_BUCKETS_ORDINAL];
+    StructField valueField = inputType.fields()[VALUE_ORDINAL];
+
+    if (!SUPPORTED_NUM_BUCKETS_TYPES.contains(numBucketsField.dataType())) {
+      throw new UnsupportedOperationException(
+          "Expected number of buckets to be tinyint, shortint or int");
+    }
+
+    // TODO - Support UUID type.
+    DataType type = valueField.dataType();
+    if (type instanceof DateType) {
+      return new BucketInt(DataTypes.DateType);
+    } else if (type instanceof ByteType
+        || type instanceof ShortType
+        || type instanceof IntegerType) {
+      return new BucketInt(DataTypes.IntegerType);
+    } else if (type instanceof LongType) {
+      return new BucketLong(DataTypes.LongType);
+    } else if (type instanceof TimestampType) {
+      return new BucketLong(DataTypes.TimestampType);
+    } else if (type instanceof DecimalType) {
+      return new BucketDecimal(((DecimalType) type).precision(), ((DecimalType) type).scale());
+    } else if (type instanceof StringType) {
+      return new BucketString();
+    } else if (type instanceof BinaryType) {
+      return new BucketBinary();
+    } else {
+      throw new UnsupportedOperationException(
+          "Expected bucketed column to be date, tinyint, smallint, int, bigint, decimal, timestamp, string, or binary");
+    }
+  }
+
+  @Override
+  public String description() {
+    return name()
+        + "(numBuckets, col) - Call Iceberg's bucket transform\n"
+        + "  numBuckets :: number of buckets to divide the rows into, e.g. bucket(100, 34) -> 79 (must be a tinyint, smallint, or int)\n"
+        + "  col :: column to bucket (must be a date, integer, long, timestamp, decimal, string, or binary)";
+  }
+
+  @Override
+  public String name() {
+    return "bucket";
+  }
+
+  public abstract static class BucketBase implements ScalarFunction<Integer> {
+    @Override
+    public String name() {
+      return "bucket";
+    }
+
+    @Override
+    public DataType resultType() {
+      return DataTypes.IntegerType;
+    }
+  }
+
+  // Used for both int and date - tinyint and smallint are upcasted to int by Spark.
+  public static class BucketInt extends BucketBase {
+    private final DataType sqlType;
+
+    // magic method used in codegen
+    public static int invoke(int numBuckets, int value) {
+      return (BucketUtil.hashInteger(value) & Integer.MAX_VALUE) % numBuckets;
```

Comment:
Can you move the mask and mod operation to a static helper method? I don't think that we want to copy that in each implementation.

Reply:
Yeah I believe so. I originally had the subclasses using a tempted type in an earlier design, but too many things needed "static" copies due to the inability to use static with templated types.

I encountered some issues in a previous design with Spark not sending the generic `produceResult` definition to executors when it was in the base class, but I think with a static function that shouldnt happen.

Will try to make it work tonight now that there's no template type.

Reply:
You could also put this in the base class.

Reply:
Made an `apply` method in the base class.